src/pymor/algorithms/chol_qr.py

```diff
@@ -0,0 +1,159 @@
+        repeat the algorithm until the check passes or `maxiter` is reached.
+    check_finite
+        This argument is passed down to |SciPy linalg| functions. Disabling may give a
+        performance gain, but may result in problems (crashes, non-termination) if the
+        inputs do contain infinities or NaNs.
+    copy
+        If `True`, create a copy of `A` instead of modifying `A` in-place.
+
+    Returns
+    -------
+    Q
+        The orthonormalized |VectorArray|.
+    R
+        The upper-triangular/trapezoidal matrix (if `compute_R` is `True`).
+    """
+    assert 0 <= offset <= len(A)
+    assert 0 < maxiter
+    assert orth_tol is None or 0 < orth_tol
+
+    if copy:
+        A = A.copy()
+
+    if len(A) == 0 or offset == len(A):
+        return A, np.eye(len(A))
+
+    logger = getLogger('pymor.algorithms.chol_qr.shifted_chol_qr')
+
+    if maxiter == 1:
+        logger.warning('Single iteration shifted CholeskyQR can lead to poor orthogonality!')
+
+    B, X = np.split(A[offset:].inner(A, product=product), [offset], axis=1)
+    B = B.conj()
+
+    dtype = np.promote_types(X.dtype, np.float32)
+    eps = np.finfo(dtype).eps
+    if dtype == np.float32:
+        xtrmm, xtrtri = spla.blas.strmm, spla.lapack.strtri
+    elif dtype == np.float64:
+        xtrmm, xtrtri = spla.blas.dtrmm, spla.lapack.dtrtri
+    elif dtype == np.complex64:
+        xtrmm, xtrtri = spla.blas.ctrmm, spla.lapack.ctrtri
+    elif dtype == np.complex128:
+        xtrmm, xtrtri = spla.blas.ztrmm, spla.lapack.ztrtri
+
+    iter = 1
+    shift = None
+    while iter <= maxiter:
+        with logger.block(f'Iteration {iter}'):
+            # This will compute the Cholesky factor of the lower right block
+            # and keep applying shifts if it breaks down.
+            while True:
+                try:
+                    Rx = spla.cholesky(X, overwrite_a=True, check_finite=check_finite)
+                    break
+                except spla.LinAlgError:
+                    pass
+                logger.warning('Cholesky factorization broke down! Matrix is ill-conditioned.')
+                if shift is None:
+                    m, n = A.dim, len(A[offset:])
+                    if product is None:
+                        shift = m*n+n*(n+1)
+                        XX = X
+                    else:
+                        from pymor.algorithms.eigs import eigs
+                        shift = 2*m*np.sqrt(m*n)+n*(n+1)*np.sqrt(np.abs(eigs(product, k=1)[0][0]))
+                        XX = A[offset:].gramian(product=product)
+                    shift *= 11*eps*spla.norm(XX, ord=2, check_finite=check_finite)
+
+                logger.info(f'Applying shift: {shift}')
+                idx = range(len(X))
+                X[idx, idx] += shift
+
+            # orthogonalize
+            Rinv = xtrtri(Rx)[0].T
+            A_todo = A[:offset].lincomb(-Rinv@B) + A[offset:].lincomb(Rinv)
```

Comment:
This formula seems wrong. In particular,

```python
from pymor.basic import *
from pymor.algorithms.chol_qr import shifted_chol_qr
U = NumpyVectorSpace(100).random(3)
gram_schmidt(U, copy=False)
U.append(U.random(2))
Q, R = shifted_chol_qr(U, offset=3, maxiter=1)
```

yields a non-orthogonal `Q`. (The problem is mitigated with a larger number of iterations.) I'm not sure if we should try to fix and merge this today, also taking into account that the PR author is not available. What do you think @lbalicki, @pmli?

Reply:
That part looks ok. The Cholesky should be applied to `X - B @ B.T` instead of `X`:

```python
Rx = spla.cholesky(X - B @ B.T, overwrite_a=True, check_finite=check_finite)
```

Reply:
@artpelling, @pymor/main, who wants to work on this? The PR is assigned to me. I don't really now the theory, however ..

Reply:
Although this PR was stagnating a bit, I think it is quite mature. I was mainly waiting for decisions on the interface discussion. @maxbindhak told me that he found a bug. I am happy to move forward on this. I've put it on the agenda for Wednesday.

Reply:
> yields a non-orthogonal `Q`. (The problem is mitigated with a larger number of iterations.)

Didn't you get a warning about the non-orthogonality? i think it is by design of the algorithm that a single iteration might not yield good orthogonality.

Reply:
> > yields a non-orthogonal `Q`. (The problem is mitigated with a larger number of iterations.)
> 
> Didn't you get a warning about the non-orthogonality? i think it is by design of the algorithm that a single iteration might not yield good orthogonality.

If I recall correctly, the change I mentioned above reduces the number of iterations for that example to 1.